internal/engine/wazevo/engine.go

```diff
@@ -227,27 +228,88 @@ func (e *engine) compileModule(ctx context.Context, module *wasm.Module, listene
 	needCallTrampoline := callTrampolineIslandSize > 0
 	var callTrampolineIslandOffsets []int // Holds the offsets of trampoline islands.
 
-	for i := range module.CodeSection {
-		if wazevoapi.DeterministicCompilationVerifierEnabled {
-			i = wazevoapi.DeterministicCompilationVerifierGetRandomizedLocalFunctionIndex(ctx, i)
-		}
+	type CompiledLocalFuncResult struct {
+		Body             []byte
+		RelsPerFunc      []backend.RelocationInfo
+		IDX              wasm.Index
+		SourceOffsetInfo []backend.SourceOffsetInfo
+	}
+
+	compiledFuncs := make([]CompiledLocalFuncResult, len(module.CodeSection))
+
+	workers := experimental.GetCompilationWorkers(ctx)
+
+	wg := sync.WaitGroup{}
+	wg.Add(workers)
+
+	ctx, cancel := context.WithCancelCause(ctx)
+	defer cancel(nil)
+
+	var count atomic.Uint32
+
+	for range workers {
+		go func() {
+			defer wg.Done()
+
+			ssaBuilder := ssa.NewBuilder()
+			machine := newMachine()
+			fe := frontend.NewFrontendCompiler(module, ssaBuilder, &cm.offsets, ensureTermination, withListener, needSourceInfo)
+
+			for {
+				// Get a stable reference to the outer context.
+				ctx := ctx
+				if err := ctx.Err(); err != nil {
+					// Compilation canceled!
+					return
+				}
+
+				i := int(count.Add(1)) - 1
+				if i >= len(module.CodeSection) {
+					return
+				}
+
+				if wazevoapi.DeterministicCompilationVerifierEnabled {
+					i = indexes[i]
+				}
 
-		fidx := wasm.Index(i + importedFns)
+				fidx := wasm.Index(i + importedFns)
 
-		if wazevoapi.NeedFunctionNameInContext {
-			def := module.FunctionDefinition(fidx)
-			name := def.DebugName()
-			if len(def.ExportNames()) > 0 {
-				name = def.ExportNames()[0]
+				if wazevoapi.NeedFunctionNameInContext {
+					def := module.FunctionDefinition(fidx)
+					name := def.DebugName()
+					if len(def.ExportNames()) > 0 {
+						name = def.ExportNames()[0]
+					}
+					ctx = wazevoapi.SetCurrentFunctionName(ctx, i, fmt.Sprintf("[%d/%d]%s", i, len(module.CodeSection)-1, name))
+				}
+
+				be := backend.NewCompiler(ctx, machine, ssaBuilder)
```

Comment:
why `be` is created per function? I believe this could make the single worker case **much** slower and memory consuming. could you fix it to be reused?

Reply:
could you check and see if there's really no impact on the regular compilation by doing the profile for like compiling zig's Stdlib or whatever huge binary as in the Standard Library Tests workflow